### PR TITLE
Upgrade React Redux and Redux Form

### PR DESCRIPTION
**Pull request by jsnajdr**
Renovate bot keeps force-pushing over my changes in #32129, so I'm creating a new PR with all the human interventions that are needed.

This will be a tough upgrade, as there are two breaking changes:

**Refs to connected components:**
`connect` used to require `withRef: true` flag if a ref was expected to be forwarded to the inner component. If it wasn't present, the ref would be assigned the outer, connected component.

In v7, the `withRef` flag was renamed to `forwardRef`. That's not hard to change. But there's additional gotcha: if `forwardRef` is not present, the ref passed to the outer component will be ignored. It will not be set to anything, not even the outer component. That can cause subtle issues that can be discovered only by manual testing.

I needed to change the following components:
- Image Editor Canvas in media image editor. Ref is used to extract the edited image from the `canvas` element.
- Zoninator extension autocomplete. Ref used to forward <kbd>Up</kbd> and <kbd>Down</kbd> key events inside search box to an autocomplete dropdown.
- Notifications: scrolling the notification list as you move around with keyboard and open notifications. Here we have instances of the tricky cases with no `withRef`: things were silently failing.

**No `store` in legacy React context:**
At several places, we extract the `store` from legacy React context (defining `contextTypes` and using `this.context`. This no longer works, because `store` is now in new React context (with provider and consumer components).

The following places need to be updated:
- the `RootChild` component that transfers the context to a different React mount tree
- TinyMCE wrappers in Classic Editor and in Woo Store that forward the store to TinyMCE and its plugins
- Guided Tours
- Signup

**Tests:**
I also needed to update a few tests that used `react-dom/test-utils`, because they started failing for some reason with the new `react-redux`. I ended up rewriting them with Enzyme, its full mount capability and the easy-to-use component wrappers. With hindsight, all that was needed was probably wrapping the renders with `act( ... )`, but I think that the Enzyme rewrite is still definitely a clear step forward.

### CODE REVIEW

**Comment by sgomes** [comment]
> but I think that the Enzyme rewrite is still definitely a clear step forward.

Oh, I thought we were trying to drop Enzyme as a dependency and rewriting things with ReactDOM test utils wherever possible?

File: apps/notifications/src/panel/templates/index.jsx [hunk 1]
```diff
@@ -1,12 +1,14 @@
 /**
- * @module templates/index
+ * External dependencies
```
Comment (sgomes) [inline apps/notifications/src/panel/templates/index.jsx]: Are the changes in this file related to this PR? I'm ok with keeping them, but I just want to make sure.

File: client/components/tinymce/index.jsx [hunk 2]
```diff
@@ -589,4 +587,15 @@ export default class extends React.Component {
 			</div>
 		);
 	}
+
+	render() {
+		return (
+			<ReactReduxContext.Consumer>
```
Comment (sgomes) [inline client/components/tinymce/index.jsx]: That's a clever, minimal fix, that doesn't entail too much rewriting 👍

File: client/jetpack-connect/help-button.jsx [hunk 3]
```diff
@@ -1,43 +1,40 @@
...
  * Internal dependencies
  */
 import LoggedOutFormLinkItem from 'components/logged-out-form/link-item';
-import { localize } from 'i18n-calypso';
 import { recordTracksEvent } from 'state/analytics/actions';
 
-export class JetpackConnectHelpButton extends PureComponent {
-	static propTypes = { label: PropTypes.string };
+export default function JetpackConnectHelpButton( { label } ) {
+	const dispatch = useDispatch();
```
Comment (sgomes) [inline client/jetpack-connect/help-button.jsx]: \o/

**Review by sgomes — COMMENTED** [review]
Code-wise, this LGTM. Left some very minor comments. Will try to run some tests now.

**Review by sgomes — APPROVED** [review]
I played around with several things you touched in the live branch and couldn't find any issues 👍 

The E2E tests should also add some measure of safety.

LGTM!

File: apps/notifications/src/panel/templates/index.jsx (hunk 1 above)
Comment (jsnajdr) [reply apps/notifications/src/panel/templates/index.jsx]: Some changes are ESLint fixes, this one related to missing internal/external docblock comments. And I also removed the `@module` docblock pragma -- we don't use it anywhere else.

**Comment by jsnajdr** [comment]
> Oh, I thought we were trying to drop Enzyme as a dependency and rewriting things with ReactDOM test utils wherever possible?

I'm not so sure about Enzyme tests that use the full React mount with JSDOM -- Enzyme's `mount` uses `react-dom/test-utils` internally and provides a convenient wrapper around it.

Some benefits of the Enzyme wrapper:
- good API for finding components and elements inside the rendered tree (`.find`, `.findWhere`), as opposed to functions with silly names like `scryRenderedDOMComponentsWithClass` (originates from [some internal FB libraries](https://stackoverflow.com/questions/33742899/where-does-reacts-scryrendereddomcomponentswithclass-method-name-come-from))
- there is a library of Jest `expect` extensions in [jest-enzyme](https://www.npmjs.com/package/jest-enzyme) that allows writing test assertions in a high-level and idiomatic way.

I don't want to go too deep into the React testing debate, as I have limited knowledge and no strong opinions about it. I mainly wanted to fix the tests that got broken during the `react-redux` debate. And also remove Chai and Sinon from some of the updated tests.

**Comment by sgomes** [comment]
> I don't want to go too deep into the React testing debate, as I have limited knowledge and no strong opinions about it. I mainly wanted to fix the tests that got broken during the `react-redux` debate. And also remove Chai and Sinon from some of the updated tests.

Sounds good to me 👍 Sorry if I derailed the conversation, but I was curious 🙂

**Pull request merged**